tracing/src/macros.rs

```diff
@@ -878,9 +878,63 @@ macro_rules! event {
 /// }
 /// ```
 ///
+/// # Alternatives
+///
+/// `enabled!` queries subscribers with [`Metadata`] where
+/// [`is_event`] and [`is_span`] both return `false`. Alternatively,
+/// use [`enabled_event`] or [`enabled_span`] to ensure one of these
+/// returns true.
+///
+///
+/// [`Metadata`]: crate::Metadata
+/// [`is_event`]: crate::Metadata::is_event
+/// [`is_span`]: crate::Metadata::is_span
+/// [`enabled_event`]: crate::enabled_event
+/// [`enabled_span`]: crate::enabled_span
+///
 #[macro_export]
 macro_rules! enabled {
-    (target: $target:expr, $lvl:expr, { $($fields:tt)* } )=> ({
+    ($($rest:tt)*)=> (
+        $crate::_enabled!(kind: { $crate::metadata::Kind::HINT }, $($rest)*)
+    )
+}
+
+/// The same as [`enabled`], but queries subscribers with `Metadata` where
+/// [`Metadata::is_event`] returns true.
+///
+/// See also [`enabled_span`].
+///
+/// [`enabled`]: crate::enabled
+/// [`Metadata::is_event`]: crate::Metadata::is_event
+/// [`enabled_span`]: crate::enabled_span
+///
+#[macro_export]
+macro_rules! enabled_event {
+    ($($rest:tt)*)=> (
+        $crate::_enabled!(kind: $crate::metadata::Kind::EVENT.hint(), $($rest)*)
+    )
+}
+
+/// The same as [`enabled`], but queries subscribers with `Metadata` where
+/// [`Metadata::is_span`] returns true.
+///
+/// See also [`enabled_event`].
+///
+/// [`enabled`]: crate::enabled
+/// [`Metadata::is_span`]: crate::Metadata::is_span
+/// [`enabled_event`]: crate::enabled_event
+///
+#[macro_export]
+macro_rules! enabled_span {
+    ($($rest:tt)*)=> (
+        $crate::_enabled!(kind: $crate::metadata::Kind::SPAN.hint(), $($rest)*)
+    )
+}
+
+#[doc(hidden)]
+#[macro_export]
+macro_rules! _enabled {
+    (kind: $kind:expr, target: $target:expr, $lvl:expr, { $($fields:tt)* } )=> ({
```

Comment:
hmm...i don't love that this is factored out into a separate macro; i'd prefer to avoid doing that when possible. what do you think about just making all the various cases with `kind`s be part of the main `enabled!` macro, and have the event and span versions just call into that, instead?

i'm open to being convinced otherwise, here!

Reply:
I got this working, but some macro stuff I don't understand made it so I had to have some weird ordering of branches...